File: go/strictcli/coverage.go
package strictcli

import (
	"bufio"
	"bytes"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"
)

// recordCoverage appends a coverage record for the resolved command path.
// Each Test() or Call() invocation writes one JSONL line to a per-process
// shard file. The shard counter increments per-write for uniqueness.
func (a *App) recordCoverage(cmdPath string) {
	if !a.testCoverage || a.coverageShardFmt == "" {
		return
	}
	path := fmt.Sprintf(a.coverageShardFmt, a.coverageCounter)
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return
	}
	f, err := os.OpenFile(path, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o644)
	if err != nil {
		return
	}
	defer f.Close()
	data, _ := json.Marshal(map[string]string{"command": cmdPath})
	f.Write(data)
	f.Write([]byte("\n"))
}

// collectAllCommandPaths enumerates all non-deprecated leaf command paths
// as dot-separated strings (e.g. "deploy", "infra.deploy").
func (a *App) collectAllCommandPaths() map[string]bool {
	paths := make(map[string]bool)
	for name := range a.commands {
		paths[name] = true
	}
	var walkGroup func(grp *Group, prefix []string)
	walkGroup = func(grp *Group, prefix []string) {
		for cmdName := range grp.Commands {
			paths[strings.Join(append(prefix, cmdName), ".")] = true
		}
		for subName, subGrp := range grp.Groups {
			walkGroup(subGrp, append(prefix, subName))
		}
	}
	for groupName, grp := range a.groups {
		walkGroup(grp, []string{groupName})
	}
	return paths
}

// testCoverageProvider is the built-in check provider for cli-test-coverage.
// Auto-registered when WithTestCoverage() is used.
//
// The verdict is derived from committed state: the covered set is the union of
// the committed manifest (.strictcli/test-coverage.json) and any per-process
// shard files merged from .strictcli/coverage/. Every live registered command
// path (minus the injected check command) must be present in that union to
// pass; otherwise the check fails naming each uncovered command.
//
// Because the verdict reads the committed manifest, it is deterministic on every
// machine -- a machine that never ran the suite (no local shards) still gets a
// stable verdict from the committed manifest alone. Both the coverage dir and
// the manifest path are anchored to the App's construction-time cwd, so the
// check evaluated from a foreign cwd reads the app's own repo state.
//
// The manifest is rewritten as the monotonic union of its prior contents and
// the freshly merged shards, but ONLY when that content actually changes -- a
// pure check must not dirty a byte-identical file. Accepted staleness: deleting
// a test leaves its command covered in the manifest until the manifest is
// deliberately regenerated (e.g. by removing it and re-running the suite),
// because the union never removes a command.
func (a *App) testCoverageProvider() []CheckSpec {
	impl := func(ctx CheckContext, reporter *ErrorReporter) CheckOutcome {
		coverageDir := a.coverageDir
		manifestPath := a.coverageManifestPath

		covered := make(map[string]bool)

		// Seed from the committed manifest -- this is what makes the verdict
		// deterministic on machines that never ran the suite.
		if manifestPath != "" {
			if raw, err := os.ReadFile(manifestPath); err == nil {
				var manifestCmds []string
				if err := json.Unmarshal(raw, &manifestCmds); err == nil {
					for _, cmd := range manifestCmds {
						covered[cmd] = true
					}
				}
			}
		}

		// Merge shards (optional freshness input)
		entries, err := os.ReadDir(coverageDir)
		if err == nil {
			for _, entry := range entries {
				if entry.IsDir() || !strings.HasSuffix(entry.Name(), ".jsonl") {
					continue
				}
				fpath := filepath.Join(coverageDir, entry.Name())
				f, err := os.Open(fpath)
				if err != nil {
					continue
				}
				scanner := bufio.NewScanner(f)
				for scanner.Scan() {
					line := strings.TrimSpace(scanner.Text())
					if line == "" {
						continue
					}
					var record map[string]string
					if err := json.Unmarshal([]byte(line), &record); err != nil {
						continue
					}
					if cmd, ok := record["command"]; ok {
						covered[cmd] = true
					}
				}
				f.Close()
			}
		}

		// Rewrite the manifest as the monotonic union, but only when the content
		// actually changes (keeps a pure check from dirtying a byte-identical
		// file).
		if manifestPath != "" && len(covered) > 0 {
			manifest := make([]string, 0, len(covered))
			for cmd := range covered {
				manifest = append(manifest, cmd)
			}
			sort.Strings(manifest)
			data, _ := json.MarshalIndent(manifest, "", "  ")
			newContent := append(data, '\n')
			existing, _ := os.ReadFile(manifestPath)
			if !bytes.Equal(existing, newContent) {
				os.MkdirAll(filepath.Dir(manifestPath), 0o755)
				os.WriteFile(manifestPath, newContent, 0o644)
			}
		}

		// Compare against command surface (exclude the framework-injected
		// check command -- it is not a user command)
		allCommands := a.collectAllCommandPaths()
		delete(allCommands, "check")
		var uncovered []string
		for cmd := range allCommands {
			if !covered[cmd] {
				uncovered = append(uncovered, cmd)
			}
		}
		sort.Strings(uncovered)

		if len(uncovered) > 0 {
			for _, cmd := range uncovered {
				reporter.Error(fmt.Sprintf("no test coverage for command: %s", cmd))
			}
			return reporter.Found(fmt.Sprintf("%d command(s) with zero test coverage", len(uncovered)))
		}
		return reporter.Passed(fmt.Sprintf("all %d commands have test coverage", len(allCommands)))
	}

	return []CheckSpec{
		NewErrorCheckSpec(CheckSpecMeta{
			Name:         "cli-test-coverage",
			Tags:         []string{"test"},
			Severity:     "error",
			Fast:         true,
			Pure:         true,
			NeedsNetwork: false,
			DependsOn:    []string{},
		}, impl),
	}
}

File: typescript/src/checks/coverage.ts
/**
 * CLI test-coverage instrumentation: per-process shard files recording which
 * commands app.test() exercised, plus the built-in cli-test-coverage check
 * provider that merges the committed manifest and shard files into the
 * covered set and compares it against the app's full command surface.
 *
 * Parity sources: go/strictcli/coverage.go with Python _record_coverage /
 * _collect_all_command_paths / _test_coverage_provider as the divergence
 * ground truth.
 */

import {
	appendFileSync,
	mkdirSync,
	readdirSync,
	readFileSync,
	statSync,
	writeFileSync,
} from "node:fs";
import { dirname, join, resolve } from "node:path";
import type { AppImpl, GroupImpl } from "../app.js";
import {
	errTestCoverageCannotCreateDir,
	RegistrationError,
} from "../errors.js";
import type { CheckSpec } from "./provider.js";
import { errorCheckSpec } from "./provider.js";

/**
 * Enables test-coverage instrumentation on an app. Anchors the coverage root
 * to the cwd AT CONSTRUCTION TIME (both the recorder and the check provider
 * use these absolute paths, so tests which chdir still record into the repo
 * and a check evaluated from a foreign cwd reads the app's own repo state),
 * creates the shard directory eagerly (a failure here is a hard construction
 * error), and registers the built-in cli-test-coverage provider.
 */
export function initTestCoverage(app: AppImpl): void {
	app.coverageDir = resolve(join(".strictcli", "coverage"));
	app.coverageManifestPath = resolve(join(".strictcli", "test-coverage.json"));
	// One shard per process (append semantics); uniqueness across concurrent
	// writers comes from the PID, so there is no per-write shard counter.
	app.coverageShardPath = join(app.coverageDir, `${process.pid}.jsonl`);
	try {
		mkdirSync(app.coverageDir, { recursive: true });
	} catch (e) {
		throw new RegistrationError(
			errTestCoverageCannotCreateDir((e as Error).message),
		);
	}
	app.registerCheckProvider(testCoverageProvider(app));
}

/**
 * Appends a coverage record for the resolved command path. Each test() or
 * call() invocation appends one JSONL line to the per-process shard file.
 */
export function recordCoverage(app: AppImpl, cmdPath: string): void {
	if (!app.testCoverage || app.coverageShardPath === undefined) {
		return;
	}
	mkdirSync(dirname(app.coverageShardPath), { recursive: true });
	appendFileSync(
		app.coverageShardPath,
		`${JSON.stringify({ command: cmdPath })}\n`,
	);
}

/**
 * Enumerates all non-deprecated leaf command paths as dot-separated strings
 * (e.g. "deploy", "infra.deploy").
 */
export function collectAllCommandPaths(app: AppImpl): Set<string> {
	const paths = new Set<string>();
	for (const name of app.commands.keys()) {
		paths.add(name);
	}
	const walkGroup = (grp: GroupImpl, prefix: readonly string[]): void => {
		for (const cmdName of grp.commands.keys()) {
			paths.add([...prefix, cmdName].join("."));
		}
		for (const [subName, subGrp] of grp.groups) {
			walkGroup(subGrp, [...prefix, subName]);
		}
	};
	for (const [groupName, grp] of app.groups) {
		walkGroup(grp, [groupName]);
	}
	return paths;
}

/** True when the path exists and is a regular file. */
function isFile(path: string): boolean {
	try {
		return statSync(path).isFile();
	} catch {
		return false;
	}
}

/** Names of shard files (*.jsonl) in the coverage dir, or [] if unreadable. */
function shardNames(coverageDir: string): string[] {
	try {
		return readdirSync(coverageDir).filter((f) => f.endsWith(".jsonl"));
	} catch {
		return [];
	}
}

/**
 * The built-in check provider for cli-test-coverage, auto-registered when
 * the app enables testCoverage. The verdict is derived from committed state:
 * the covered set is the union of the committed manifest
 * (.strictcli/test-coverage.json) and any per-process shard files merged from
 * .strictcli/coverage/. Every live registered command path (minus the
 * injected check command) must be present in that union to pass; otherwise
 * the check fails naming each uncovered command.
 *
 * Because the verdict reads the committed manifest, it is deterministic on
 * every machine -- a machine that never ran the suite (no local shards) still
 * gets a stable verdict from the committed manifest alone.
 *
 * The manifest is rewritten as the monotonic union of its prior contents and
 * the freshly merged shards, but ONLY when that content actually changes -- a
 * pure check must not dirty a byte-identical file. Accepted staleness:
 * deleting a test leaves its command covered in the manifest until the
 * manifest is deliberately regenerated, because the union never removes a
 * command.
 */
export function testCoverageProvider(app: AppImpl): () => CheckSpec[] {
	return () => [
		errorCheckSpec({
			name: "cli-test-coverage",
			tags: ["test"],
			fast: true,
			pure: true,
			needsNetwork: false,
			dependsOn: [],
			impl: (_ctx, reporter) => {
				const coverageDir = app.coverageDir ?? "";
				const manifestPath = app.coverageManifestPath ?? "";

				// Subject-matter gating (the sanctioned skip class, mirroring
				// project-type gating): when the anchored coverage root holds
				// NEITHER a committed manifest NOR any shard files, this is not
				// the app's own development tree -- e.g. an installed app running
				// its checks from a foreign project's cwd. Report a visible SKIP
				// instead of failing with the app's entire command surface listed
				// as uncovered. When EITHER exists, behavior is unchanged: a
				// partial manifest still fails honestly, and an empty-manifest
				// file present still means "coverage configured but empty" = fail
				// listing all.
				const manifestExists = manifestPath !== "" && isFile(manifestPath);
				const shards = coverageDir !== "" ? shardNames(coverageDir) : [];
				if (!manifestExists && shards.length === 0) {
					const anchor =
						manifestPath !== "" ? dirname(manifestPath) : coverageDir;
					return reporter.skipped(
						`no coverage state at ${anchor} -- cli-test-coverage applies to the app's own development tree`,
					);
				}

				const covered = new Set<string>();

				// Seed from the committed manifest -- this is what makes the
				// verdict deterministic on machines that never ran the suite.
				if (manifestExists) {
					try {
						const data: unknown = JSON.parse(
							readFileSync(manifestPath, "utf8"),
						);
						if (Array.isArray(data)) {
							for (const c of data) {
								if (typeof c === "string") {
									covered.add(c);
								}
							}
						}
					} catch {
						// Unreadable/invalid manifest contributes nothing.
					}
				}

				// Merge shards (optional freshness input).
				for (const fname of shards) {
					const content = readFileSync(join(coverageDir, fname), "utf8");
					for (const rawLine of content.split("\n")) {
						const line = rawLine.trim();
						if (line === "") {
							continue;
						}
						const entry = JSON.parse(line) as Record<string, unknown>;
						if (typeof entry.command === "string") {
							covered.add(entry.command);
						}
					}
				}

				// Rewrite the manifest as the monotonic union, but only when the
				// content actually changes (keeps a pure check from dirtying a
				// byte-identical file).
				if (manifestPath !== "" && covered.size > 0) {
					const manifest = [...covered].sort();
					const newContent = `${JSON.stringify(manifest, null, 2)}\n`;
					let existing: string | undefined;
					try {
						existing = readFileSync(manifestPath, "utf8");
					} catch {
						existing = undefined;
					}
					if (existing !== newContent) {
						mkdirSync(dirname(manifestPath), { recursive: true });
						writeFileSync(manifestPath, newContent);
					}
				}

				// Compare against the command surface (exclude the
				// framework-injected check command -- it is not a user command).
				const allCommands = collectAllCommandPaths(app);
				allCommands.delete("check");
				const uncovered = [...allCommands]
					.filter((cmd) => !covered.has(cmd))
					.sort();

				if (uncovered.length > 0) {
					for (const cmd of uncovered) {
						reporter.error(`no test coverage for command: ${cmd}`);
					}
					return reporter.found(
						`${uncovered.length} command(s) with zero test coverage`,
					);
				}
				return reporter.passed(
					`all ${allCommands.size} commands have test coverage`,
				);
			},
		}),
	];
}
